src/main/scala/fi/aalto/cs/apluscourses/intellij/actions/ReplAction.scala

```diff
@@ -0,0 +1,113 @@
+    }
+  }
+
+  def setConfigurationConditionally(@NotNull project: Project,
+                                    @NotNull module: Module,
+                                    @NotNull configuration: ScalaConsoleRunConfiguration): Unit = {
+    if (ReplConfigurationFormModel.showREPLConfigWindow) {
+      val configModel = new ReplConfigurationFormModel(project,
+        getModuleWorkDir(module), module.getName)
+
+      createAndShowReplConfigurationDialog(configModel)
+
+      if (!configModel.isStartRepl) {
+        throw new CancelReplStartException
+      }
+
+      val changedModuleName = configModel.getTargetModuleName
+      val changedModule = ModuleManager.getInstance(project).findModuleByName(changedModuleName)
+      val changedWorkDir = configModel.getModuleWorkingDirectory
+      setCustomConfigurationFields(configuration, changedWorkDir, changedModuleName, changedModule)
+    } else {
+      setCustomConfigurationFields(configuration, getModuleWorkDir(module), module.getName, module)
+    }
+  }
+
+  private def createAndShowReplConfigurationDialog(@NotNull configModel: ReplConfigurationFormModel):
+  ReplConfigurationDialog = {
+    val configForm = new ReplConfigurationForm(configModel)
+    val configDialog = new ReplConfigurationDialog
+    configDialog.setReplConfigurationForm(configForm)
+    configDialog.setVisible(true)
+    configDialog
+  }
+
+  /**
+   * Method that sets working directory and module of the REPL it was started from. Works for REPL
+   * triggered on files or folders within the module scope.
+   *
+   * @param e an [[AnActionEvent]] with payload.
+   */
+  def customDoRunAction(@NotNull e: AnActionEvent): Unit = {
+    val dataContext = e.getDataContext
+    val project = CommonDataKeys.PROJECT.getData(dataContext)
+    //  virtual file is working for both: files and folders
+    val targetFileOrFolder = CommonDataKeys.VIRTUAL_FILE.getData(dataContext)
+
+    if (project == null || checkFileOrFolderIsNull(targetFileOrFolder)) return // scalastyle:ignore
+
+    //  get target module & workDir
+    val module = ModuleUtilCore.findModuleForFile(targetFileOrFolder, project)
+
+    val runManagerEx = RunManagerEx.getInstanceEx(project)
+    val configurationType = getMyConfigurationType
+    val settings = runManagerEx.getConfigurationSettingsList(configurationType).asScala
+
+    //  choose the configuration to run based on the condition if this a new configuration of not
+    val setting = settings.headOption.getOrElse {
+      val factory = configurationType.getConfigurationFactories.head
+      runManagerEx.createConfiguration(s"REPL in ${module.getName}", factory)
+    }
+
+    val configuration = setting.getConfiguration.asInstanceOf[ScalaConsoleRunConfiguration]
+    try {
+      setConfigurationConditionally(project, module, configuration)
+    } catch {
+      case _: CancelReplStartException => return // scalastyle:ignore
+    }
+
+    RunConsoleAction.runExisting(setting, runManagerEx, project)
```

Comment:
I would personally avoid using exceptions for control flow. I have added a PR with an alternative possibility, but feel free to disagree of course :smile: